### PR TITLE
Program Admin View now displays applications sorted by their Submit Time (descending), also applies to JSON and CSV exports. #8743

### LINKED ISSUE #8743: New applications are not listed from newest to oldest in program admin view
**Describe the bug**
This is a known issue uncovered by the Seattle Team. If there is an existing issue that I didn't see this can be deleted or closed. 

New submitted applications are not sorting to the top of the list in the Civiform program admin view. Here is a screen shot of them currently out of order. No search or date filters were applied to result in the order shown. 

screenshot

**To Reproduce**
Steps to reproduce the behavior:/

go to the civiform program admin and view applications for the Weatherization program. Applications are not showing up in order of newest to oldest. 

**Expected behavior**

**Screenshots**
PDF linked above. 

**Desktop (please complete the following information):**

**Smartphone (please complete the following information):**

**Additional context**

### PR DESCRIPTION
**Pull request by lagosr-google**
### Description

This change refactors the pagination spec objects into a object oriented hierarchy, so that we can implement 2 different approaches to pagination, and the different kind of sorting the application uses today.

The new pagination spec hierarchy:

- **BasePaginationSpec**: the abstract interface definition of a pagination spec.

And its subclass implementations:

- **PageNumberPaginationSpec**: sorts by submit time or by id, and can provide random access to any page in the results.
- **SubmitTimeSequentialAccessPaginationSpec**: sorts by submit time, and provides sequential access to pages.
- **RowIdSequentialAccessPaginationSpec**: sorts by ID, and provides sequential access to pages.

Pagination comes in 2 flavors: random access and sequential access. 

- **Random access** allows retrieval of any page in the result at any time.  performance suffers the higher the page number being accessed, because the query must read every object up to the page being served to determine which objects to serve.
- **Sequential access** allows for constant time in query performance, but requires that each page in the result be accessed in sequence, since the pagination object requires the id of the last object read in order to serve the next page.

Each spec has a recommended use, sequential access is good for API methods, and export methods; and random access is good for user views.

This PR fixes the original bug, changing the sort order of the application admin view to use submitTime and id as the sort keys.  It also makes it so that the CSV / JSON export will have the same sort order.

API results will still return items sorted by object ID.

- Verified that the view shows the corrected order for applications
- Verified that the fix also shows the submit time order in the CSV and JSON downloads.
- Verified that API keys view sorts by API key ID. 
- Removed the old pagination objects and all references.

Moving to object oriented model for the pagination spec allows the implementation of how to affect the query's sort order and where clause to be implemented once in the spec object (rather than multiple times outside of the object), and provides a cleaner implementation in the controller and repository classes.

Finally since there were now 4 files related to PaginationSpec's I thought it merited having them in their own directory, as a follow up i will move the other 2 pagination related classes into that directory ASAP.  (PaginationResult and PaginationInfo)

#### General

Read the full guidelines for PRs [here](https://github.com/civiform/civiform/wiki/Technical-contribution-guide#creating-a-pull-request)

- [x] Added the correct label: < feature | enhancement | bug | under-development | dependencies | infrastructure | ignore-for-release | database >
- [x] Assigned to a specific person, `civiform/developers`, or a [more specific round-robin list](https://github.com/civiform/civiform/wiki/Technical-contribution-guide#adding-reviewers)
- [x] Added an additional reviewer from outside your organization as FYI (if the primary reviewer is in the same organization as you)
- [x] Removed the release notes section if the title is sufficient for the release notes description, or put more details in that section.
- [x] Created unit and/or browser tests which fail without the change (if possible)
- [x] Performed manual testing (Chrome and Firefox if it includes front-end changes)
- [x] Extended the README / documentation, if necessary. For user-facing features, consider updating [the user docs](https://github.com/civiform/docs). For "under-the-hood" changes or things more relevant to developers, consider updating [the dev wiki](https://github.com/civiform/civiform/wiki).

### Issue(s) this completes

Fixes #8743

### CODE REVIEW

**Review by avaleske — COMMENTED** [review]
- We need to add an index on submit_time, right?
- Can we add a test for the SubmitTimePaginationSpec when the submit time is set?
- Can we add a test for the RowIdPaginationSpec with a page size that isn't the max?

Also:
- The PR title is what will show up in release notes, so naming it using the same convention as google3 would be clearer.
- Be sure to add another reviewer that isn't at google as the "fyi reviewer". This ensures that changes are supported by more than one org on the project.

File: server/app/services/pagination/SubmitTimePaginationSpec.java [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+    if (this.currentSubmitTime.equals(Instant.MAX)) {
+      return query.where().lt("id", this.currentRowId).query();
+    }
+    return query
+        .where()
+        .or()
+        .and()
+        .eq("submitTime", Date.from(this.currentSubmitTime))
+        .lt("id", this.currentRowId)
+        .lt("submitTime", this.currentRowId)
```
[the remark this answers is not in the document]
Comment (lagosr-google) [reply server/app/services/pagination/SubmitTimePaginationSpec.java:58]: the formatted is fighting me here -- if i can't get it to format the way i'd like i will add a comment here .

and yes i forgot the endOr and the endAnd here .. and there is a bug in the actual parameters .. this was a part of the code we dont have testing for .. so i'm adding a test to get coverage and will get it to behave properly.

File: server/app/services/pagination/SubmitTimePaginationSpec.java [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+  protected <T> Query<T> applyOrderBy(Query<T> query) {
+    return query.orderBy("submitTime desc, id desc");
+  }
+
+  @Override
+  protected <T> Query<T> maybeApplyWhere(Query<T> query) {
+    // Date.from(Instant.MAX) is not supported, (overflows). If that is current
+    // submit time in the spec, then skip setting a submitTime in the where
+    // clause(), since all values in the database should be before the
+    // Instant.MAX Date.
```
[the remark this answers is not in the document]
Comment (lagosr-google) [reply server/app/services/pagination/SubmitTimePaginationSpec.java:48]: java instant doesn't believe there is time beyond MAX.

File: server/app/services/PaginationInfo.java [hunk 3]
```diff
@@ -2,6 +2,8 @@
 
 import com.google.common.collect.ImmutableList;
 
+// TODO(lagosr): Move this class into the pagination subpackage.
```
[the remark this answers is not in the document]
Comment (lagosr-google) [reply server/app/services/PaginationInfo.java:5]: https://github.com/civiform/civiform/issues/8841 added to track the moe -- will update the TODO with the BUG id

File: server/app/services/pagination/BasePaginationSpec.java [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+ * performance degrades as higher pages are queried for. Usually used in an user-facing interface.
+ *
+ * <p>Sequential paging: requires knowledge of the last row returned by the previous page to
+ * retrieve the next page. This spec has constant performance for any page in the range, but can
+ * only access pages in sequence. Usually used for export functions and API retrieve methods.
+ *
+ * <p>Implementations of this interface are: PageNumberPaginationSpec, RowIdPaginationSpec, and
+ * SubmitTimePaginationSpec.
+ *
+ * <p>PageNumberPaginationSpec: sorts by submit time, is a random access paging spec. Recommended
```
[the remark this answers is not in the document]
Comment (lagosr-google) [reply server/app/services/pagination/BasePaginationSpec.java:21]: done

File: server/app/services/pagination/BasePaginationSpec.java [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+   */
+  protected abstract <T> Query<T> applyOrderBy(Query<T> query);
+
+  /** optionally implemented in the subclass if implementing a sequential access paging spec */
+  protected <T> Query<T> maybeApplyWhere(Query<T> query) {
+    return query;
+  }
+
+  /** optionally implemented in the subclass if implementing a random access paging spec. */
+  protected <T> Query<T> maybeApplySetFirstRow(Query<T> query) {
```
[the remark this answers is not in the document]
Comment (lagosr-google) [reply server/app/services/pagination/BasePaginationSpec.java:84]: given that its totally up to the subclass to do the right thing and we dont have any enforcement, the current way is more honest -- the added comments and renaming of the subclasses makes it more clear how to implement a new flavor of the paginators.  I'd like to leave as is.

File: server/app/services/pagination/BasePaginationSpec.java [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+   * clause.
+   */
+  protected abstract <T> Query<T> applyOrderBy(Query<T> query);
+
+  /** optionally implemented in the subclass if implementing a sequential access paging spec */
+  protected <T> Query<T> maybeApplyWhere(Query<T> query) {
+    return query;
+  }
+
+  /** optionally implemented in the subclass if implementing a random access paging spec. */
```
[the remark this answers is not in the document]
Comment (lagosr-google) [reply server/app/services/pagination/BasePaginationSpec.java]: done (updated)

File: server/app/services/pagination/SubmitTimePaginationSpec.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (lagosr-google) [reply server/app/services/pagination/SubmitTimePaginationSpec.java]: there are 2 new tests... one for this conditional and one for rthe non MAX PerId Paginator

File: server/test/repository/ProgramRepositoryTest.java [hunk 7]
```diff
@@ -846,6 +849,116 @@ public void getApplicationsForAllProgramVersions_sortedBySubmitDate() {
...
+    ApplicationModel applicationThree =
+        resourceCreator.insertActiveApplication(applicantThree, nextVersion);
+
+    /* Set the submit date such that the results come back in this order: 2, 1, 3 */
+    applicationThree.setSubmitTimeToNow();
+    applicationThree.save();
+    applicationOne.setSubmitTimeToNow();
+    applicationOne.save();
+    applicationTwo.setSubmitTimeToNow();
+    applicationTwo.save();
```
Comment (avaleske) [inline server/test/repository/ProgramRepositoryTest.java:879]: Do these come through as different times?

If so, it might be good to have a test that also tests when a set of multiple applications come through at the same time, with the page boundary in the middle of that set, to make sure the ID ordering is working correctly too.

**Review by avaleske — APPROVED** [review]
Thank you for this!

I had one additional thought about another test but otherwise it LGTM!

**Comment by avaleske** [comment]
I'm also guessing errorprone is mad at you.

File: server/app/services/pagination/BasePaginationSpec.java [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+
+import io.ebean.Query;
+
+/**
+ * Abstracts different kind of pagination strategies.
+ *
+ * <p>PaginationSpecs come in 2 major flavors: random acess paging and sequential paging.
+ *
+ * <p>Random Access Paging: requires no knowledge of the last row that was displayed in a previous
+ * page. This spec requires reading from storage all the rows prior to the page being returned,
```
Comment (gwendolyngoetz) [inline server/app/services/pagination/BasePaginationSpec.java:11]: (question) What do you mean by "requires reading from storage all the rows"? Does our code not leverage `limit` and `offset` in postgres?

File: server/test/repository/ProgramRepositoryTest.java (hunk 7 above)
Comment (lagosr-google) [reply server/test/repository/ProgramRepositoryTest.java:879]: yes, these sets different times. 

I didn't add a test for the same specific time, because that test exists all over the code right now .. just about every test that exists right now neglects to set a submit time , and so you end up with the applications sorted by ID ..

File: server/app/services/pagination/BasePaginationSpec.java (hunk 8 above)
Comment (lagosr-google) [reply server/app/services/pagination/BasePaginationSpec.java:11]: lol - -i had written a very nice long reply .. and then I hit "resolve conversation" rather than "add comment".  *sigh*

In order to do an offset, the database has to count the number of rows up to the offset.  An Index can't help here because you don't know what the value of the indexed columns will be at an offset.

The amount of work done to count storage rows up to the offset grows linearly as the offset count increases (increasing page #'s).

So limit does help not read beyond a certain point, but offset only helps not SERVE rows we don't care about, but on the db side you still had to do "work" that grows linearly.

Comment (gwendolyngoetz) [reply server/app/services/pagination/BasePaginationSpec.java:11]: Gotcha, the way it's worded sounded like it was all being loaded into memory in the application.

**Review by avaleske — APPROVED** [review]
